### PR TITLE
Make Galaxy (almost) packageable/installable

### PR DESCRIPTION
**Pull request by natefoo**
# 2019 Update

This won't be entirely ready for `pip install galaxy` but does what's needed to quickly finish it up. The plan is for the [galaxy-web-apps package](https://github.com/galaxyproject/galaxy/tree/dev/packages/web_apps/) package essentially a full install of Galaxy with its web stack, and the [galaxy metapackage](https://github.com/galaxyproject/galaxy/tree/dev/packages/meta/) will be pinned to stable releases. Because the packagification of Galaxy began just before updating this PR happened, the method originally used here (one big package) is no longer valid, and the metapackage isn't working yet. That will happen in a subsequent PR.

This PR includes some important useful pieces:
- `GalaxyConfiguration.config_dir` is `dirname(config_file)` (or `.` if there is no config file)
- `GalaxyConfiguration.data_dir` is `./data`
- No assumption that `cwd` is `galaxy_root`
- No assumption that configs are in `./config/`
- Sample configs are no longer required
- Galaxy now finds templates, static, and tools via symlinks in the galaxy lib dir (but these are not packageable yet)
- Sample configs moved to `lib/galaxy/config/sample`

The exception to the changes above: if running from `cwd = galaxy_root`, then `config_dir` and `data_dir` are `<cwd>/config` and `<cwd>/database` as before.

## Updated TODO list for the next PR
- [ ] Finish the metapackage
- [ ] Update `galaxy-config` for YAML configs
- [ ] Add `galaxy-uwsgi` or `galaxy -m uwsgi`
- [ ] Package `static/`
- [ ] Other directories: `external_services/`, `tool-data/`, ...
- [ ] Create a `galaxy-set-metadata` command to replace the tempfile method currently used
- [ ] Move `galaxy-manage-db` to galaxy-app after #8293 I think
- [ ] Fix tool migrations
- [ ] Other things I'm not thinking of

# Original PR

Building on #428, this attempts to make Galaxy installable via the standard Pythonic mechanism. To use:
1. Make a virtualenv and activate it
2. `PYTHONPATH= pip install --upgrade pip wheel`
3. `make wheels`
4. `PYTHONPATH= pip install -i https://wheels.galaxyproject.org/simple/ 'dist/galaxy-16.7.dev0-py2-none-any.whl[postgresql]'`
5. Create a directory somewhere, copy `config/galaxy.ini.sample` to it as `galaxy.ini`
6. `galaxy-paster serve galaxy.ini`

Alternatively:
1. Make a virtualenv and activate it
2. `PYTHONPATH= pip install --upgrade pip`
3. `make client`
4. `PYTHONPATH= pip install -i https://wheels.galaxyproject.org/simple '.[postgresql]'` (add `-e` for editable/develop mode)
5. As above
## TODO list
- [x] Figure out what to do with `lib/galaxy/datatypes/converters/`, `lib/galaxy/datatypes/set_meatadata_tool.xml`, etc. (just use `__file__` rather than assuming cwd is galaxy_root?)
- [x] Fix `galaxy-main`
- [x] Make a `galaxy-config` to create a Galaxy config directory and relevant contents
- [x] Add a `config_dir` option to galaxy.ini
- [x] Move `shed_tool_conf.xml` to its own config variable
- [x] Do not require `shed_*` configs to exist (they can be created as necessary)
- [x] Figure out what to do with all the mutable stuff (compiled templates, sqlite db, etc.)
- [x] ~~Install custom pip~~ not needed
- [x] Install dependencies
- [x] Figure out what to do with `tools/`
- [x] Distribute `templates/` and make it findable by Galaxy
- [x] Distribute `static/` and make it findable by Galaxy
- [ ] Other directories: `external_services/`, `tool-data/`, ...
- [x] Clean up the root directory and move packages to the top level
- [ ] Create a `galaxy-set-metadata` command to replace the tempfile method currently used
- [ ] Fix tool migrations
- [ ] Other things I'm not thinking of

### CODE REVIEW

**Comment by mr-c** [comment]
:100: !!

**Comment by martenson** [comment]
api tests on Jenkins report

```
Traceback (most recent call last):
  File "./scripts/manage_db.py", line 11, in <module>
    from migrate.versioning.shell import main
ImportError: No module named migrate.versioning.shell
```

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
this needs to be rebased or dev merged into it before jenkins can test

**Comment by natefoo** [comment]
@martenson Waiting for the tests to pass on #428 first.

File: setup.py
Comment (jmchilton) [inline setup.py]: Datatypes depends on models so it shouldn't live in galaxy-lib in my opinion.

**Comment by jmchilton** [comment]
What made you decide what to put in `galaxy-app` and what to put into `galaxy`? I feel like we should just figure out galaxy-lib first - and probably after this effort? After that I'd like to aim toward a design of:
- galaxy-lib (minimal required dependencies)
- galaxy-models (depends on sqlalchemy and galaxy-lib)
- galaxy-app (galaxy-models but no web stuff - someday it would be nice to run a webless galaxy process - e.g. job or workflow handlers using just this)
- galaxy-web (galaxy-app + controllers)

I would start from the bottom and introduce one or so a year as we refactor around these new distinctions? There are a think both practical and abstract design goals with each of these distinctions that I could go into in depth. None of this feels like something that should be rushed though.

**Comment by natefoo** [comment]
> What made you decide what to put in `galaxy-app` and what to put into `galaxy`?

"This seems like a web package", "This doesn't". Which is why it's almost certainly wrong.

So for now perhaps we should just have `galaxy-lib` and `galaxy`.

**Comment by nsoranzo** [comment]
It would be nice to make the final tool migration (xref. #636) before this PR is merged.

**Comment by natefoo** [comment]
Per conversation with @jmchilton we are going to add a `config_dir` option to galaxy.ini so that we can stop assuming the location of everything is relative to the current working directory. Is there a sensible default for this? I suggest the directory containing galaxy.ini.

**Comment by jmchilton** [comment]
I've never sat down and gone through this - but now that I have I stand by what I said the other day and even more so - this needing to broken into pieces 😅. This is too sprawling to ever get in as is IMO and has diverged really hard - and so many of the little pieces could have been great to have in for years.

File: Makefile
Comment (jmchilton) [inline Makefile]: This file is gone right?

File: lib/galaxy/util/__init__.py
Comment (jmchilton) [inline lib/galaxy/util/__init__.py:1708]: I'd prefer you move this into a new file. But maybe track that for after we merge this.

File: lib/tool_shed/galaxy_install/migrate/check.py
Comment (jmchilton) [inline lib/tool_shed/galaxy_install/migrate/check.py]: Note: this will be deleted 6 years from now and never fixed 😆

File: packages/meta/galaxy/project_galaxy_util.py
Comment (jmchilton) [inline packages/meta/galaxy/project_galaxy_util.py]: File name needs to be changed I think.

File: packages/meta/galaxy/project_galaxy_util.py
Comment (jmchilton) [inline packages/meta/galaxy/project_galaxy_util.py]: We switched this to ``galaxy-committers@lists.galaxyproject.org``

File: packages/meta/setup.py
Comment (jmchilton) [inline packages/meta/setup.py:38]: This belong in galaxy-app after https://github.com/galaxyproject/galaxy/pull/8293 I think.

**Review by jmchilton — APPROVED** [review]

File: lib/galaxy/util/__init__.py
Comment (natefoo) [reply lib/galaxy/util/__init__.py:1708]: Where would you prefer?

File: packages/meta/galaxy/project_galaxy_util.py
Comment (natefoo) [reply packages/meta/galaxy/project_galaxy_util.py]: And here I missed the chance to harass you about claiming authorship of all of Galaxy's code

File: packages/meta/setup.py
Comment (natefoo) [reply packages/meta/setup.py:38]: Added to the TODO

**Comment by nsoranzo** [comment]
Wooot!

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py:124]: @natefoo `os.path.dirname(os.getcwd())` is the parent of the current directory, so I think this may be wrong if `self.config_file` is None.
Also this is set to a different value on line 134 if `running_from_source` is True.

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py:933]: @natefoo While trying to fix CI testing for BioBlend I noticed that:
- `path` may be a relative, e.g. `./config/shed_tool_conf.xml`, because it comes from `resolve_path(path, self.root)` in `_parse_config_file_options()` , where `self.root` is `.`
- `self.shed_tool_conf` is an absolute path, e.g. `/srv/galaxy/config/shed_tool_conf.xml`, because it comes from `resolve_path(kwargs.get('shed_tool_conf', 'shed_tool_conf.xml'), self.mutable_config_dir)` where `self.mutable_config_dir` is an absolute path.

Making `self.root` absolute seems a good way to solve this:
```diff
diff --git a/lib/galaxy/config/__init__.py b/lib/galaxy/config/__init__.py
index 92ab49fccd..aa1edd1f5f 100644
--- a/lib/galaxy/config/__init__.py
+++ b/lib/galaxy/config/__init__.py
@@ -100,7 +100,7 @@ def resolve_path(path, root):
 
 
 def find_root(kwargs):
-    root = kwargs.get('root_dir', '.')
+    root = os.path.abspath(kwargs.get('root_dir', '.'))
     return root
 
 
```

What do you think?

File: lib/galaxy/tools/data_manager/manager.py
Comment (nsoranzo) [inline lib/galaxy/tools/data_manager/manager.py:43]: @natefoo @ic4f I don't think `self.app.config.shed_data_manager_config_file_set` and `self.app.config.data_manager_config_file_set` are set anywhere.

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py:157]: Probably it's not a bug, but reusing `defaults` variable name as loop variable is surely confusing.

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py:171]: Here too.

File: lib/galaxy/tools/data_manager/manager.py
Comment (natefoo) [reply lib/galaxy/tools/data_manager/manager.py:43]: IIRC this is dynamic, but I'll double check.

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py:157]: Yes!! I noticed it and hated it on first site.. I'll submit a quick PR (didn't want to trigger a code review over such a minor thing - but since you mention this.. so thanks :)

File: lib/galaxy/tools/data_manager/manager.py
Comment (natefoo) [reply lib/galaxy/tools/data_manager/manager.py:43]: Yeah, they are set by `_parse_config_file_options()`

File: lib/galaxy/config/__init__.py
Comment (natefoo) [reply lib/galaxy/config/__init__.py:157]: @ic4f thanks!

File: lib/galaxy/tools/data_manager/manager.py
Comment (jdavcs) [reply lib/galaxy/tools/data_manager/manager.py:43]: I think they might get set here: 
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L162
Until this: https://github.com/galaxyproject/planemo/pull/950 was merged, that's where  `data_manager_config_file_set` was set (was passed with kwargs as empty string). 
That said, for future consideration, I think that block has too many special cases, which makes it not intuitive. Even a comment explaining the 'why' would help, I think.

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py:157]: @nsoranzo How about defaults >> vals? 
```
for var, vals in defaults.items():
    if config_kwargs.get(var, None) is not None:
        path = config_kwargs.get(var)
        setattr(self, var + '_set', True)
    else:
        for val in vals:
            if os.path.exists(resolve_path(val, self.root)):
                path = val 
                break
        else:
            path = vals[-1]
        setattr(self, var + '_set', False)
    setattr(self, var, resolve_path(path, self.root))
```

Comment (nsoranzo) [reply lib/galaxy/config/__init__.py:157]: Maybe `values`/`value` to make it a bit more different from `var`.

File: lib/galaxy/tools/toolbox/base.py
Comment (martenson) [inline lib/galaxy/tools/toolbox/base.py:166]: fresh on dev this is called twice with `config_filename=/Users/marten/devel/git/temp/galaxy/config/migrated_tools_conf.xml` but once it has `getattr(self.app.config, opt)` equal to `/Users/marten/devel/git/temp/galaxy/config/shed_tool_conf.xml`.

so the `_dynamic_tool_confs` end up being 

```python
[{
'tool_path': '/Users/marten/devel/git/temp/galaxy/database/shed_tools',
 'create': '<?xml version="1.0"?>\n<toolbox tool_path="/Users/marten/devel/git/temp/galaxy/database/shed_tools">\n</toolbox>\n', 
'config_filename': '/Users/marten/devel/git/temp/galaxy/config/migrated_tools_conf.xml', 
'config_elems': []}]
```
 which seems wrong?

Comment (natefoo) [reply lib/galaxy/tools/toolbox/base.py:166]: Can you create an issue and assign me?

Comment (martenson) [reply lib/galaxy/tools/toolbox/base.py:166]: The issue is here: https://github.com/galaxyproject/galaxy/issues/8484 and you are assigned to it.

Comment (natefoo) [reply lib/galaxy/tools/toolbox/base.py:166]: Ah, didn't realize this was the same thing, thanks.

**Pull request merged**
